File: 146-LruCache/146-LruCache.cpp
// Last updated: 4/3/2026, 2:01:58 PM
class LRUCache {
public:
    class Node {
    public:
        int _key;
        int _value;
        Node* next;
        Node* prev;

        Node(int _k, int _val) {
            _key = _k;
            _value = _val;
        }
    };
    Node* head = new Node(-1, -1);
    Node* tail = new Node(-1, -1);
    int cap;
    unordered_map<int, Node*> mpp;
    LRUCache(int capacity) {
        head->next = tail;
        tail->prev = head;
        cap = capacity;
    }
    void addNode(Node* newNode) {
        newNode->next = head->next;
        head->next->prev = newNode;
        newNode->prev = head;
        head->next = newNode;
    }
    void deleteNode(Node* dltNode) {
        Node* prevNode = dltNode->prev;
        Node* nextNode = dltNode->next;

        prevNode->next = dltNode->next;
        nextNode->prev = dltNode->prev;
    }
    int get(int key) {
        if (mpp.find(key) == mpp.end())
            return -1;
        
        Node* temp = mpp[key];
        int value = temp->_value;
        deleteNode(temp);
        addNode(temp);
        return value;
    }

    void put(int key, int value) {
        // Already exists
        if(mpp.find(key) != mpp.end()) {
            Node* temp = mpp[key];
            deleteNode(temp);
            mpp.erase(key);
            delete temp;
        }
        if(mpp.size() == cap) {
            Node* temp = tail->prev;
            deleteNode(temp);
            mpp.erase(temp->_key);
            delete temp;
        }
        Node* newNode = new Node(key, value);
        addNode(newNode);
        mpp[key] = newNode;
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: 146-LruCache/146-LruCache.java
// Last updated: 4/3/2026, 2:02:01 PM
class LRUCache {
    class Node {
        Node prev;
        Node next;
        int key;
        int value;

        Node(int _key, int _value) {
            key = _key;
            value = _value;
        }
    }

    private void deleteNode(Node node) {
        mpp.remove(node.key);
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

    private void insertAfterHead(Node node) {
        mpp.put(node.key, node);
        node.next = head.next;
        node.next.prev = node;
        head.next = node;
        node.prev = head;
    }

    Node head = new Node(-1, -1);
    Node tail = new Node(-1, -1);
    Map<Integer, Node> mpp = new HashMap();
    int capacity;

    public LRUCache(int cap) {
        capacity = cap;
        head.next = tail;
        tail.prev = head;
    }

    public int get(int key) {
        if (mpp.containsKey(key)) {
            Node node = mpp.get(key);
            deleteNode(node);
            insertAfterHead(node);
            return node.value;
        }
        return -1;
    }

    public void put(int key, int value) {
        if (mpp.containsKey(key)) {
            deleteNode(mpp.get(key));
        }
        if (mpp.size() == capacity) {
            deleteNode(tail.prev);
        }
        insertAfterHead(new Node(key, value));
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */